// @thirdweb-dev/react-native/dist/evm/styles/textVariants.js
export const textVariants = {
    defaults: {
        fontFamily: undefined,
        fontSize: 24,
        fontWeight: "700",
        lineHeight: 32,
        color: "textPrimary",
    },
    header: {
        fontFamily: undefined,
        fontSize: 24,
        fontWeight: "600",
        lineHeight: 32,
        color: "textPrimary",
    },
    headerBold: {
        fontFamily: undefined,
        fontSize: 24,
        fontWeight: "700",
        lineHeight: 32,
        color: "textPrimary",
    },
    subHeader: {
        fontFamily: undefined,
        fontSize: 16,
        fontWeight: "500",
        lineHeight: 24,
        color: "textSecondary",
    },
    link: {
        fontFamily: undefined,
        fontWeight: "500",
        fontSize: 14,
        lineHeight: 24,
        color: "linkPrimary",
        letterSpacing: -0.02,
    },
    bodyLarge: {
        fontFamily: undefined,
        color: "textPrimary",
        textAlign: "center",
        fontWeight: "600",
        fontSize: 16,
        lineHeight: 24,
    },
    bodyLargeBold: {
        fontFamily: undefined,
        color: "textPrimary",
        textAlign: "center",
        fontWeight: "700",
        fontSize: 16,
        lineHeight: 24,
    },
    bodySmall: {
        fontFamily: undefined,
        color: "textPrimary",
        textAlign: "center",
        fontWeight: "600",
        fontSize: 14,
        lineHeight: 16,
    },
    bodySmallBold: {
        fontFamily: undefined,
        color: "textPrimary",
        textAlign: "center",
        fontWeight: "700",
        fontSize: 14,
        lineHeight: 16,
    },
    bodySmallSecondary: {
        fontFamily: undefined,
        color: "textSecondary",
        textAlign: "left",
        fontWeight: "600",
        fontSize: 14,
        lineHeight: 16,
    },
    error: {
        fontFamily: undefined,
        color: "error",
        fontWeight: "500",
        fontSize: 12,
        lineHeight: 16,
        textAlign: "center",
    },
};
