// @thirdweb-dev/react-native/dist/evm/styles/theme.js
import { deepMerge } from "../types/deepPartial";
import { paletteDark, paletteLight } from "./colors";
import { textVariants } from "./textVariants";
import { createTheme } from "@shopify/restyle";
export const baseTheme = {
    spacing: {
        none: 0,
        xxs: 6,
        xs: 8,
        sm: 12,
        md: 16,
        xmd: 20,
        lg: 24,
        xl: 32,
        xxl: 48,
    },
    borderRadii: {
        xs: 4,
        sm: 6,
        md: 8,
        lg: 12,
        xl: 20,
        xxl: 32,
    },
    textVariants,
};
export const _darkTheme = createTheme({
    ...baseTheme,
    colors: {
        ...paletteDark,
    },
});
export const _lightTheme = {
    ...baseTheme,
    colors: {
        ...paletteLight,
    },
};
export const darkTheme = ({ buttonBackgroundColor = paletteDark.buttonBackgroundColor, buttonTextColor = paletteDark.buttonTextColor, buttonBorderColor = paletteLight.buttonBorderColor, colors, ...props } = _darkTheme) => {
    return deepMerge(_darkTheme, {
        colors: {
            ..._darkTheme.colors,
            ...colors,
            buttonBackgroundColor,
            buttonTextColor,
            buttonBorderColor,
        },
        ...props,
    });
};
export const lightTheme = ({ buttonBackgroundColor = paletteLight.buttonBackgroundColor, buttonTextColor = paletteLight.buttonTextColor, buttonBorderColor = paletteLight.buttonBorderColor, colors, ...props } = _lightTheme) => {
    return deepMerge(_lightTheme, {
        colors: {
            ..._lightTheme.colors,
            ...colors,
            buttonBackgroundColor,
            buttonTextColor,
            buttonBorderColor,
        },
        ...props,
    });
};
