// @thirdweb-dev/react-native/dist/evm/components/ConnectWalletFlow/ConnectingWallet/ConnectingWalletHeader.js
import { Icon } from "../../../assets/icon";
import { useGlobalTheme, useLocale, } from "../../../providers/ui-context-provider";
import Text from "../../base/Text";
import { StyleSheet, View } from "react-native";
export const ConnectWalletHeader = ({ headerText, subHeaderText, alignHeader = "center", middleContent, onClose, onBackPress, }) => {
    const l = useLocale();
    const theme = useGlobalTheme();
    subHeaderText =
        subHeaderText !== undefined
            ? subHeaderText
            : l.connecting_wallet.connecting_your_wallet;
    return (<>
      <View style={styles.header}>
        {onBackPress ? (<Icon type="back" width={14} height={14} onPress={onBackPress} color={theme.colors.iconPrimary}/>) : (<View />)}
        {middleContent ? middleContent : null}
        <Icon type="close" width={14} height={14} onPress={onClose} color={theme.colors.iconPrimary}/>
      </View>
      <View style={{ ...styles.headerContainer, alignItems: alignHeader }}>
        {headerText ? (<Text variant="header" mt="md">
            {headerText}
          </Text>) : null}
        {subHeaderText ? (<Text variant="subHeader" mt="md" textAlign="center">
            {subHeaderText}
          </Text>) : null}
      </View>
    </>);
};
const styles = StyleSheet.create({
    header: {
        display: "flex",
        flexDirection: "row",
        justifyContent: "space-between",
    },
    headerContainer: {
        display: "flex",
        flexDirection: "column",
        justifyContent: "center",
    },
});
