// @thirdweb-dev/react-native/dist/core/AsyncStorage.js
import { MMKV } from "react-native-mmkv";
const PREFIX = "__TW__";
export class AsyncLocalStorage {
    constructor(name) {
        this.name = name;
        this.asyncStorage = new MMKV();
    }
    getItem(key) {
        const item = this.asyncStorage.getString(`${PREFIX}/${this.name}/${key}`) || null;
        return Promise.resolve(item);
    }
    setItem(key, value) {
        return Promise.resolve(this.asyncStorage.set(`${PREFIX}/${this.name}/${key}`, value));
    }
    removeItem(key) {
        return Promise.resolve(this.asyncStorage.delete(`${PREFIX}/${this.name}/${key}`));
    }
}
export class LocalStorage {
    constructor(name) {
        this.name = name;
        this.asyncStorage = new MMKV();
    }
    getItem(key) {
        return this.asyncStorage.getString(`${PREFIX}/${this.name}/${key}`) || null;
    }
    setItem(key, value) {
        return this.asyncStorage.set(`${PREFIX}/${this.name}/${key}`, value);
    }
    removeItem(key) {
        return this.asyncStorage.delete(`${PREFIX}/${this.name}/${key}`);
    }
}
export class noopStorage {
    getItem(key) {
        return Promise.resolve(key);
    }
    setItem(key, value) {
        console.log(key, value);
        return Promise.resolve();
    }
    removeItem(key) {
        console.log(key);
        return Promise.resolve();
    }
}
/**
 * Returns a new instance of AsyncLocalStorage
 *
 * @param name Name to namespace the storage with
 * @returns A new instance of AsyncLocalStorage
 */
export function createAsyncLocalStorage(name) {
    return new AsyncLocalStorage(name);
}
/**
 * Returns a new instance of LocalStorage
 *
 * @param name Name to namespace the storage with
 * @returns A new instance of LocalStorage
 */
export function createSyncStorage(name) {
    return new LocalStorage(name);
}
