// @thirdweb-dev/react-native/dist/evm/wallets/wallets/wallet-connect/WalletConnect.js
import { MMKV } from "react-native-mmkv";
import { WalletConnectBase } from "./WalletConnectBase";
import { WALLETCONNECT_ICON } from "../../../assets/svgs";
const PREFIX = "__TW__WALLET__CONNECT__";
export class WalletConnect extends WalletConnectBase {
    constructor(options) {
        super(options);
        this.storage = new MMKV();
        const storedLinksStr = this.storage.getString(PREFIX + "links");
        this.links = storedLinksStr
            ? JSON.parse(storedLinksStr)
            : {
                native: "wc:",
                universal: "wc:",
            };
        this.on("disconnect", () => {
            this.storage.delete(PREFIX + "links");
        });
    }
    getMeta() {
        return { ...WalletConnect.meta, links: this.links };
    }
    setWCLinks(links) {
        this.storage.set(PREFIX + "links", JSON.stringify(links));
        this.links = links;
    }
}
WalletConnect.id = "WalletConnect";
WalletConnect.meta = {
    name: "WalletConnect",
    iconURL: WALLETCONNECT_ICON,
    links: {
        native: "wc:",
        universal: "wc:",
    },
};
